### ISSUE TITLE
[-v] Recover from setenv

### ISSUE THREAD

**Issue by whit537** [whit537/cloud-custodian#31]
In https://github.com/capitalone/cloud-custodian/pull/1326 we added `setenv` to `tox.ini`, which [frobulates](https://github.com/whit537/cloud-custodian/pull/27#issuecomment-310336307) ratchet somehow. 'Sup with that?

**Comment by whit537** [comment]
It's the `C7N_VALIDATE=true`. Validation doesn't work with Python 3.

**Comment by whit537** [comment]
```
$ .tox/py36/bin/pytest --tb=native tests/test_account.py::AccountTests -k test_service_limit_specific_check
================================ test session starts ================================
platform darwin -- Python 3.6.1, pytest-3.1.2, py-1.4.34, pluggy-0.4.0
rootdir: /Users/whit537/workbench/capitalone/cloud-custodian, inifile: tox.ini
plugins: xdist-1.17.1, cov-2.5.1
collected 18 items 

tests/test_account.py .

================================ 17 tests deselected ================================
====================== 1 passed, 17 deselected in 2.02 seconds ======================
$ C7N_VALIDATE=true .tox/py36/bin/pytest --tb=native tests/test_account.py::AccountTests -k test_service_limit_specific_check
============================= test session starts ==============================
platform darwin -- Python 3.6.1, pytest-3.1.2, py-1.4.34, pluggy-0.4.0
rootdir: /Users/whit537/workbench/capitalone/cloud-custodian, inifile: tox.ini
plugins: xdist-1.17.1, cov-2.5.1
collected 18 items

tests/test_account.py F

=================================== FAILURES ===================================
________________ AccountTests.test_service_limit_specific_check ________________
Traceback (most recent call last):
  File "/Users/whit537/workbench/capitalone/cloud-custodian/tests/test_account.py", line 178, in test_service_limit_specific_check
    session_factory=session_factory)
  File "/Users/whit537/workbench/capitalone/cloud-custodian/tests/common.py", line 99, in load_policy
    errors = schema_validate({'policies': [data]}, C7N_SCHEMA)
  File "/Users/whit537/workbench/capitalone/cloud-custodian/c7n/schema.py", line 48, in validate
    errors = list(validator.iter_errors(data))
  File "/Users/whit537/workbench/capitalone/cloud-custodian/.tox/py36/lib/python3.6/site-packages/jsonschema/validators.py", line 105, in iter_errors
    for error in errors:
  File "/Users/whit537/workbench/capitalone/cloud-custodian/.tox/py36/lib/python3.6/site-packages/jsonschema/_validators.py", line 304, in properties_draft4
    schema_path=property,
  File "/Users/whit537/workbench/capitalone/cloud-custodian/.tox/py36/lib/python3.6/site-packages/jsonschema/validators.py", line 121, in descend
    for error in self.iter_errors(instance, schema):
  File "/Users/whit537/workbench/capitalone/cloud-custodian/.tox/py36/lib/python3.6/site-packages/jsonschema/validators.py", line 105, in iter_errors
    for error in errors:
  File "/Users/whit537/workbench/capitalone/cloud-custodian/.tox/py36/lib/python3.6/site-packages/jsonschema/_validators.py", line 55, in items
    for error in validator.descend(item, items, path=index):
  File "/Users/whit537/workbench/capitalone/cloud-custodian/.tox/py36/lib/python3.6/site-packages/jsonschema/validators.py", line 121, in descend
    for error in self.iter_errors(instance, schema):
  File "/Users/whit537/workbench/capitalone/cloud-custodian/.tox/py36/lib/python3.6/site-packages/jsonschema/validators.py", line 105, in iter_errors
    for error in errors:
  File "/Users/whit537/workbench/capitalone/cloud-custodian/.tox/py36/lib/python3.6/site-packages/jsonschema/_validators.py", line 364, in anyOf_draft4
    errs = list(validator.descend(instance, subschema, schema_path=index))
  File "/Users/whit537/workbench/capitalone/cloud-custodian/.tox/py36/lib/python3.6/site-packages/jsonschema/validators.py", line 121, in descend
    for error in self.iter_errors(instance, schema):
  File "/Users/whit537/workbench/capitalone/cloud-custodian/.tox/py36/lib/python3.6/site-packages/jsonschema/validators.py", line 105, in iter_errors
    for error in errors:
  File "/Users/whit537/workbench/capitalone/cloud-custodian/.tox/py36/lib/python3.6/site-packages/jsonschema/_validators.py", line 216, in ref
    for error in validator.descend(instance, resolved):
  File "/Users/whit537/workbench/capitalone/cloud-custodian/.tox/py36/lib/python3.6/site-packages/jsonschema/validators.py", line 121, in descend
    for error in self.iter_errors(instance, schema):
  File "/Users/whit537/workbench/capitalone/cloud-custodian/.tox/py36/lib/python3.6/site-packages/jsonschema/validators.py", line 105, in iter_errors
    for error in errors:
  File "/Users/whit537/workbench/capitalone/cloud-custodian/.tox/py36/lib/python3.6/site-packages/jsonschema/_validators.py", line 333, in allOf_draft4
    for error in validator.descend(instance, subschema, schema_path=index):
  File "/Users/whit537/workbench/capitalone/cloud-custodian/.tox/py36/lib/python3.6/site-packages/jsonschema/validators.py", line 121, in descend
    for error in self.iter_errors(instance, schema):
  File "/Users/whit537/workbench/capitalone/cloud-custodian/.tox/py36/lib/python3.6/site-packages/jsonschema/validators.py", line 105, in iter_errors
    for error in errors:
  File "/Users/whit537/workbench/capitalone/cloud-custodian/.tox/py36/lib/python3.6/site-packages/jsonschema/_validators.py", line 304, in properties_draft4
    schema_path=property,
  File "/Users/whit537/workbench/capitalone/cloud-custodian/.tox/py36/lib/python3.6/site-packages/jsonschema/validators.py", line 121, in descend
    for error in self.iter_errors(instance, schema):
  File "/Users/whit537/workbench/capitalone/cloud-custodian/.tox/py36/lib/python3.6/site-packages/jsonschema/validators.py", line 105, in iter_errors
    for error in errors:
  File "/Users/whit537/workbench/capitalone/cloud-custodian/.tox/py36/lib/python3.6/site-packages/jsonschema/_validators.py", line 55, in items
    for error in validator.descend(item, items, path=index):
  File "/Users/whit537/workbench/capitalone/cloud-custodian/.tox/py36/lib/python3.6/site-packages/jsonschema/validators.py", line 121, in descend
    for error in self.iter_errors(instance, schema):
  File "/Users/whit537/workbench/capitalone/cloud-custodian/.tox/py36/lib/python3.6/site-packages/jsonschema/validators.py", line 105, in iter_errors
    for error in errors:
  File "/Users/whit537/workbench/capitalone/cloud-custodian/.tox/py36/lib/python3.6/site-packages/jsonschema/_validators.py", line 364, in anyOf_draft4
    errs = list(validator.descend(instance, subschema, schema_path=index))
  File "/Users/whit537/workbench/capitalone/cloud-custodian/.tox/py36/lib/python3.6/site-packages/jsonschema/validators.py", line 121, in descend
    for error in self.iter_errors(instance, schema):
  File "/Users/whit537/workbench/capitalone/cloud-custodian/.tox/py36/lib/python3.6/site-packages/jsonschema/validators.py", line 105, in iter_errors
    for error in errors:
  File "/Users/whit537/workbench/capitalone/cloud-custodian/.tox/py36/lib/python3.6/site-packages/jsonschema/_validators.py", line 201, in enum
    if instance not in enums:
TypeError: unhashable type: 'dict'
============================= 17 tests deselected ==============================
=================== 1 failed, 17 deselected in 0.79 seconds ====================
```

**Comment by whit537** [comment]
```
------------------------------- Captured stdout call --------------------------------
BLUE FALAFEL: account ['iam-group']
BLUE FALAFEL: service-limit ['event']
BLUE FALAFEL: service-limit ['has-inline-policy']
BLUE FALAFEL: service-limit ['has-users']
BLUE FALAFEL: service-limit ['value']
BLUE FALAFEL: {'type': 'service-limit', 'limits': ['DB security groups'], 'threshold': 1.0} dict_keys(['value', 'or', 'and', 'not', 'event', 'has-users', 'has-inline-policy'])
```

**Comment by whit537** [comment]
Hmmm ... that code is never hit under 2.7? (I did add the print in the py27 env.)

```
$ C7N_VALIDATE=true .tox/py27/bin/pytest --tb=native tests/test_account.py::AccountTests -k test_service_limit_specific_check
================================ test session starts ================================
platform darwin -- Python 2.7.10, pytest-3.1.2, py-1.4.34, pluggy-0.4.0
rootdir: /Users/whit537/workbench/capitalone/cloud-custodian, inifile: tox.ini
plugins: xdist-1.17.1, cov-2.5.1
collected 18 items 

tests/test_account.py .

================================ 17 tests deselected ================================
====================== 1 passed, 17 deselected in 2.45 seconds ======================
$
```

**Comment by whit537** [comment]
Fixed somewhere under #33.

**Comment by whit537** [comment]
I think [this log](https://github.com/whit537/cloud-custodian/pull/33#issuecomment-312658252) is relevant:

> It *is* called. :thinking:
> 
> ```
> (Pdb) l
> 199     
> 200     def enum(validator, enums, instance, schema):
> 201         print('BLUE FALAFEL:', instance, enums)
> 202         if type(instance) is dict:
> 203             import import pdb; pdb.set_trace(); import pdb; pdb.set_trace().set_trace()
> 204  ->     if instance not in enums:
> 205             yield ValidationError("%r is not one of %r" % (instance, enums))
> 206     
> 207     
> 208     def ref(validator, ref, instance, schema):
> 209         resolve = getattr(validator.resolver, "resolve", None)
> (Pdb) instance not in enums
> True
> (Pdb) instance
> {u'threshold': 1.0, u'type': u'service-limit', u'limits': [u'DB security groups']}
> (Pdb)
> ```
> 
> ----------------
> 
> Is this a bug in jsonschema? Does jsonschema support Python 3? 
> 
> ----------------
> 
> `dict` is unhashable in both!
> 
> ```
> $ python2.7 -c 'hash(dict())'
> Traceback (most recent call last):
>   File "<string>", line 1, in <module>
> TypeError: unhashable type: 'dict'
> $ python3.6 -c 'hash(dict())'
> Traceback (most recent call last):
>   File "<string>", line 1, in <module>
> TypeError: unhashable type: 'dict'
> $
> ```
> 
> ----------------
> 
> And that means that a `dict` can't be a `dict` key in both:
> 
> ```
> $ python2.7 -c '{} in {}'
> Traceback (most recent call last):
>   File "<string>", line 1, in <module>
> TypeError: unhashable type: 'dict'
> $ python3.6 -c '{} in {}'
> Traceback (most recent call last):
>   File "<string>", line 1, in <module>
> TypeError: unhashable type: 'dict'
> $ 
> ```
> 
> ----------------
> 
> Aha! `enums` is a `list` under 2.7 but a `dict_keys` under 3.6:
> 
> ```
> (Pdb) enums
> [u'and', u'value', u'event', u'metrics', u'not', u'or']
> (Pdb) 
> ```
> 
> vs.
> 
> ```
> (Pdb) enums
> dict_keys(['value', 'or', 'and', 'not', 'event', 'has-users', 'has-inline-policy'])
> (Pdb)
> ```
> 
> ----------------
> 
> Alright, so where does `enums` come from? Do we control that? I *think* we do ...
> 
> ----------------
> 
> ```
> (Pdb) pp schema
> {'anyOf': [{'$ref': '#/definitions/resources/iam-group/filters/event'},
>            {'$ref': '#/definitions/resources/iam-group/filters/has-inline-policy'},
>            {'$ref': '#/definitions/resources/iam-group/filters/has-users'},
>            {'$ref': '#/definitions/resources/iam-group/filters/value'},
>            {'$ref': '#/definitions/filters/valuekv'},
>            {'enum': dict_keys(['value', 'or', 'and', 'not', 'event', 'has-users', 'has-inline-policy'])}]}
> (Pdb)
> ```
